### PR TITLE
Remove excluded test cases based on tests.toml when generating cases_test.go

### PR DESCRIPTION
**Pull request by AbhilashJN**
#2055 

Added a way to recursively parse the `canonical-data.json` for a problem specification, and remove test cases which are to be excluded as per the `tests.toml` file for that exercise.

### CODE REVIEW

**Review by eklatzer — COMMENTED** [review]
Hey, as this is already doing some stuff mentioned in #2055 I will not start on my own implementation, but I will give you feedback about your implementation. 
* I would personally rename `parseTestsToml.go` to `toml.go` and `filterTests.go` to `filter.go`
* I do not think using a `map[string]bool{}` is needed, as we only need something like a set datastructure, which can be done by using `map[string]struct{}{}` (here a discussion: https://stackoverflow.com/questions/34018908/golang-why-dont-we-have-a-set-datastructure)
* instead of parsing to string I would use the following snippet:
```go
for scanner.Scan() {
		line := scanner.Text()
		line = strings.ReplaceAll(line, " ", "")
		if len(line) == 0 || strings.HasPrefix(line, "#") {
			continue
		}

		if strings.HasPrefix(line, "[") {
			currentCaseUUID = strings.Trim(line, "[]")
			continue
		}

		if isExcluded(line) {
			excludedList[currentCaseUUID] = struct{}{}
		}
	}

func isExcluded(line string) bool {
	parts := strings.Split(line, "=")
	key := parts[0]
	value := parts[1]
	if key == "include" {
		return value == "false"
	}
	return false
}
```

File: gen/filterTests.go
Comment (eklatzer) [inline gen/filterTests.go:29]: This part is not needed if `var s` is of type `map[string]json.RawMessage`

File: gen/gen.go
Comment (eklatzer) [inline gen/gen.go:138]: Error messages should not be capitalized: https://stackoverflow.com/questions/68792696/why-error-messages-shouldnt-end-with-a-punctuation-mark-in-go#:~:text=Error%20strings%20should%20not%20be%20capitalized%20(unless%20beginning%20with%20proper,usually%20printed%20following%20other%20context.

**Review by eklatzer — COMMENTED** [review]
Hey, thanks for the quick response. I think the `toml.go` is pretty clean now. Using `strings.TrimSpace` seems to be better than my suggestion.

File: gen/filter.go
Comment (eklatzer) [inline gen/filter.go:65]: Can we somehow improve this? It is basically a duplicate of lines 14-26.

**Comment by junedev** [comment]
@AbhilashJN @eklatzer Great work on this you two! @eklatzer You cannot officially approve this but please let me know when you are ok with the result. Then I will give it a final check and merge.

Great that there was a test case added. I was just about to suggest that. I will also create an issue for the future about adding tests for other parts of the generator.

**Comment by eklatzer** [comment]
Hey, thanks for the feedback. I don't know if we want to merge this one. I have used the code of this PR in #2279 as these two topics are strongly connected. I hope it is OK that I used the code there, if not...sorry

**Comment by junedev** [comment]
@eklatzer The issue is that if I only merge your PR, @AbhilashJN will not get the credit (e.g. reputation on the website) they deserve for this so that would be unfair. This should be merged and then your PR can iterate on it as you please. I'm sorry for the inconvenience that causes you in resolving merge conflicts etc.

**Comment by eklatzer** [comment]
OK, then I will wait for this PR to be merged. Could you review the code? I will also go through it later

**Review by junedev — APPROVED** [review]
I think there are some more minor improvements that could be done in the code here but no need to further block this from being merged. Also it does not do anything to the production code so if we find issues later when using the modified generator, we can always fix them later on. Also further improvements of the code can be made in future PRs.

I tested the full generation process with a simple example and one with sub groups and everything seems to work as expected.

**Comment by junedev** [comment]
@eklatzer Sorry, I only saw your additional comments after I clicked merge. Feel free to simply fix those things as part of your own PR.

**Comment by eklatzer** [comment]
Hey, is it OK if I create a new branch based on the current `main` and then add my changes with better commit messages than the ones I added when refactoring the test case generator? So #2279 could be closed without being merged

**Comment by junedev** [comment]
@eklatzer Whatever you prefer but be aware that we squash commits so the individual commits get lost anyway.

**Comment by junedev** [comment]
@AbhilashJN Thanks for getting this done. You should now be able to continue with #2170 after updating your branch with the latest changes from the upstream main branch.

**Comment by andrerfcsantos** [comment]
Huge thanks @AbhilashJN and @eklatzer!

This was not something trivial and you handled it really well <3

**Pull request merged**
